added prometheus exporter on /metrics path

Tall-Paul opened the pull request:
shows basic info (load, wifi strength, free memory etc) and device values

Commit:
added prometheus exporter on /metrics path

shows basic info (load, wifi strength, free memory etc) and device values

Commit:
added prometheus exporter on /metrics path

shows basic info (load, wifi strength, free memory etc) and device values

fixes from PR

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+
+
+
+    web_server.send(200, F("text/plain;"), resultString);
```
Please use `addHtml` for this as you're now creating a massive string in memory.

Tall-Paul replied:
It's not html though, I did dig into the 'standard' way you have for constructing web pages, but right at the start you output a 'text/html', I need it to be 'text/plain' and not have any html tags in the output.  I'll take another stab at it

TD-er replied:
You can look at how the JSON page is rendered.
How the header is sent is independent on how the text is sent to `TXBuffer`

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+
+String handle_metrics_value_name(const String& valName){
+    return String(valName);
+}
```
This is a rather silly function as it just copies the data.

Tall-Paul replied:
Yeah, forgot to remove these

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+}
+
+String handle_metrics_value_value(const String& valValue){
+    return String(valValue);
```
idem

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+
+
+String handle_metrics_devices(){
+    String returnString = "";
```
No need to initialize with an empty string.
The default constructor of `String` does create a valid empty string.

TD-er replied:
This one still appears to be present.
`String returnString;` is enough.
May sound strange, but adding an empty string initialization like this does really add unneeded code, including using extra RAM as it is a string and not a F() macro string.
And like I said, the default constructor of `String` is a valid empty string.

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+    String returnString = "";
+
+
+    for (taskIndex_t x = 0; x < 128 && validTaskIndex(x); x++)
```
Why not use MAX_TASKS here instead of 128?

Tall-Paul replied:
Because I didn't know about max tasks :D

TD-er replied:
Not sure if MAX_TASKS is the correct one, you can check the content of `valisTaskIndex`

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+        const bool pluginID_set         = INVALID_PLUGIN_ID != Settings.TaskDeviceNumber[x];
+         if (pluginID_set){
+             LoadTaskSettings(x);
+            int8_t spi_gpios[3] { -1, -1, -1 };
```
`spi_gpios` is not used?

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+             LoadTaskSettings(x);
+            int8_t spi_gpios[3] { -1, -1, -1 };
+            struct EventStruct TempEvent(x);
+            addEnabled(Settings.TaskDeviceEnabled[x]  && validDeviceIndex(DeviceIndex));
```
The function addEnabled does add HTML code. Is that what you want?

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+                        if (validPluginID_fullcheck(Settings.TaskDeviceNumber[x]))
+                        {
+                              String valName = String(F("valuename_")) + x + '_' + varNr;
+                              String valValue = String(F("value_")) + x + '_' + varNr;
```
`valName` and `valValue` are not used.

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,190 @@
+            
+         }
+    }
+    return returnString;
```
Again, you're creating a massive string, while you can directly send it to the client using `addHtml`

Tall-Paul committed:
fixes from PR

Tall-Paul committed:
Merge branch 'feature-add-prometheus-exporter' of https://github.com/Tall-Paul/ESPEasy into feature-add-prometheus-exporter

Tall-Paul commented:
Okay, @TD-er think I've fixed all those issues

stream to page instead of constructing a string
removed unused functions / variables
changed that loop to remove the '128'
Also added a check for enabled devices  (rather than all of them)

Tall-Paul committed:
Merge branch 'mega' into feature-add-prometheus-exporter

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -61,130 +61,115 @@
-    resultString += "\n";
+    addHtml(F("# HELP espeasy_free_ram device amount of RAM free in Bytes\n"));
+    addHtml(F("# TYPE espeasy_free_ram gauge\n"));
+    addHtml("espeasy_free_ram ");
```
Please wrap these in F() macro

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -61,130 +61,115 @@
+    addHtml(F("# TYPE espeasy_free_ram gauge\n"));
+    addHtml("espeasy_free_ram ");
+    addHtml(getValue(LabelType::FREE_MEM));
+    addHtml("\n");
```
`addHtml('\n');`
`\n` is a single char, and way less computational expensive compared to a string

Tall-Paul replied:
Ha, can you tell I've never written any C/C++ in anger? :D

TD-er replied:
"Anger"?? relax :)

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -1,9 +1,9 @@
+# include "../WebServer/WebServer.h"
+# include "../WebServer/HTML_wrappers.h"
+# include "../WebServer/Markup.h"
+# include "../WebServer/Markup_Buttons.h" 
```
I'm quite sure not all of these includes are needed here.

TD-er commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -61,130 +61,115 @@
+
     int result;
     String resultString;
     resultString.reserve(1000);
```
This `resultString` can also go as it is no longer needed and with the reserve it still consumes a kB of RAM

Tall-Paul committed:
more PR fixes

Tall-Paul committed:
Merge branch 'feature-add-prometheus-exporter' of https://github.com/Tall-Paul/ESPEasy into feature-add-prometheus-exporter

Tall-Paul commented:
think that should resolve all of those, let me know if this is okay, I can flatten my branch to get rid of all these....less than optimal attempts

TD-er commented:
> think that should resolve all of those, let me know if this is okay, I can flatten my branch to get rid of all these....less than optimal attempts

No need to squash the commits, It is merged as a single commit.

TD-er commented:
There is another important part missing in this PR... the documentation.
What is a prometheus exporter and why should one want to use it?

Tall-Paul commented:
sigh, just realised that (because I'm copying how the json page does it) I'm getting a content type of 'text/html' rather than 'text/plain'.  Any tips for setting a content header before I go trawling through documentation?

TD-er commented:
> sigh, just realised that (because I'm copying how the json page does it) I'm getting a content type of 'text/html' rather than 'text/plain'. Any tips for setting a content header before I go trawling through documentation?

I will make a commit for you (on this PR) where I change the following function:

```c++
void Web_StreamingBuffer::sendHeaderBlocking(bool json, const String& origin) {
```

And add:
```c++
  void startStream(const String& content_type, const String& origin);
```

TD-er commented:
@Tall-Paul Since I will push it to your branch, can you confirm you're not working on a new commit?
Not sure how experienced you are with Git, but it would be a waste of time if either of our commits got lost ;)

TD-er commented:
![image](https://user-images.githubusercontent.com/3751318/127003188-538229b7-4433-4c1b-b8bb-f13be39d462d.png)
This is how it will be after my changes

TD-er committed:
[Prometheus] Add text/plain as header

TD-er commented:
By the way, does it have to use the URL `/metrics` ?
Why not `/prometheus` ? (and thus also rename the file in the WebServer directory)

Tall-Paul commented:
/metrics is the default url for prometheus to scrape.  You can change it, but unless there's a reason for doing so it might as well go with the default.  

see: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

I haven't done any further changes locally since my last push, so no worries about overwriting anything on my branch :)

Will provide some documentation in the next couple of days.

Tall-Paul committed:
Merge branch 'letscontrolit:mega' into feature-add-prometheus-exporter

Tall-Paul committed:
Added prometheus documentation

TD-er commented on src/src/WebServer/ToolsPage.cpp:
```diff
@@ -79,6 +79,8 @@ void handle_tools() {
 
   addWideButtonPlusDescription(F("json"),        F("Show JSON"),    F("Open JSON output"));
 
+  addWideButtonPlusDescription(F("metrics"),        F("Show Metrics"),    F("Open Prometheus Metrics"));
```
This also needs the `#ifdef WEBSERVER_METRICS` wrapped

TD-er commented on docs/source/Tools/Tools.rst:
```diff
@@ -485,6 +485,33 @@ New default value since 2021-06-20: unchecked
+
+This allows easy connection via prometheus to grafana for graphing, as in the screenshot below:
+
+.. image:: images/Grafana.png
```
Better rename this image name to `PrometheusGrafana.png` to make it more clear what item this image belongs to.

Tall-Paul committed:
Added ifdef, updated image name as per PR comments

Tall-Paul committed:
removed unused string definition

tonhuisman commented on src/src/DataStructs/Web_StreamingBuffer.h:
```diff
@@ -76,8 +80,9 @@ class Web_StreamingBuffer {
-                          const String& origin = "");
+  void sendHeaderBlocking(bool          allowOriginAll,
+                          const String& content_type,
+                          const String& origin);
```
We're loosing the default `= ""` on `origin` here, is that intentional?

TD-er replied:
Yep, it is a private function and only called from one point.
So better not to leave default parameters if they aren't used optionally.

tonhuisman commented on src/src/WebServer/Metrics.cpp:
```diff
@@ -0,0 +1,121 @@
+#endif // ifdef ESP32
+
+
+void handle_metrics() {
```
In `Metrics.h` these functions are conditionally compiled (`#ifdef WEBSERVER_METRICS`), shouldn't that be here as well, or does the linker take care of leaving it out of the final binary? (I'd suggest to add the `#ifdef` here too)

TD-er replied:
Ah missed that one?
The linker will remove unused functions (as long as they are not part of a class), so it won't matter here but it is good practice to have it wrapped too.
